### DISCUSSION TITLE
Calculation of derivatives of a eigenmode

### DISCUSSION THREAD

**Discussion by HelgeGehring** [kinnala/scikit-fem#955] [Q&A] [answered]
How would I apply an operator like
![image](https://user-images.githubusercontent.com/42973196/200420441-f396d819-5367-449c-9607-eb4a42c263df.png)
on the efield I calculate in https://github.com/HelgeGehring/waveguidemodes/blob/main/waveguidemodes/mode_solver.py ?

I've tried to calculate it like
```python
def calculate_hfield(basis, xs, beta):
    xs = xs.astype(complex)

    @BilinearForm(dtype=np.complex64)
    def aform(e_t, e_z, v_t, v_z, w):
        return (-1j * beta * e_t[1] + e_z.grad[1]) * v_t[0] \
               + (1j * beta * e_t[0] - e_z.grad[0]) * v_t[1] \
               + e_t.curl * v_z

    a_operator = aform.assemble(basis)

    @BilinearForm(dtype=np.complex64)
    def bform(e_t, e_z, v_t, v_z, w):
        return dot(e_t, v_t) + e_z * v_z

    b_operator = bform.assemble(basis)

    return scipy.sparse.linalg.spsolve(b_operator, a_operator @ xs)
```
When I use this, the result looks like
![image](https://user-images.githubusercontent.com/42973196/200420771-f1638e9a-0572-46d8-86a0-abb57b550c7d.png)

The x-component of the result looks fine, the y-component seems wrong and should look more like
![image](https://user-images.githubusercontent.com/42973196/200421362-6353ba37-2e05-47e9-a9c2-21a2669b2553.png)
and I really don't know what is going on in z-direction.
Is this approach to apply the operator the right one? Or how would I apply it?

**Comment by kinnala** [comment, marked as answer]
What are you plotting there? The magnitude of the complex number?

**Reply by HelgeGehring** [reply]
I'm plotting there the real part, here the imaginary part:
![image](https://user-images.githubusercontent.com/42973196/200487677-0d6f0546-6fe9-4b83-a8d6-7d573e46c23a.png)

**Reply by HelgeGehring** [reply]
Here how the Z compoent shoul look like (magnitude)
![image](https://user-images.githubusercontent.com/42973196/200488244-2b83d662-88c2-4a96-806d-561b54e38664.png)
(I've just realized, that the program I use as a reference seems not to show the magnitude as it also shows negative values)

**Reply by HelgeGehring** [reply]
How would I plot the magnitude of the field?
`plot_mode(basis, np.abs(xbs))` seems wrong to me
![image](https://user-images.githubusercontent.com/42973196/200488935-c23721ac-8bc6-47c9-ac70-a8090c1a0bbf.png)

**Reply by HelgeGehring** [reply]
Okay, I probably should calculate the magnitude after splitting the basis, right?

I.e.
```
    (et, et_basis), (ez, ez_basis) = basis.split(mode)

    plot_basis = et_basis.with_element(ElementVector(ElementTriP0()))
    et_xy = plot_basis.project(et_basis.interpolate(et))
    (et_x, et_x_basis), (et_y, et_y_basis) = plot_basis.split(et_xy)
```
But it seems that basis.project discards the imaginary part, there we would probably also need an adjustment to allow complex fields.
~~One thing which confuses me also is: might it be that et_x and et_y are somewhere exchanged? For the E-field I see the component I expect in the et_y component in the plot for the et_x component and vice versa~~ I'll get tomorrow better reference images, there seems to be an error in the package I use to compare the results to

**Reply by HelgeGehring** [reply]
Okay, I just realized that I had one factor wrong in my calculation - Hy seems now to work fine if I take `e_t[0]` instead of `-e_t[0]` - I'll search further if I have somewhere else a sign wrong.

Would you have an idea why the z-component of the calculated H-field looks so wrong?

**Reply by HelgeGehring** [reply]
Fixed it with the signs, the problem was that as I use the version of slepc which supports complex eigenvectors, getEigenpair(i, xr, xi) returns the eigenvector completely (real and imag) in xr and sets xi to zero. I'm now only using real(xr) and it seems to work fine :)

But still, I don't understand what is going on in H_z, the high-frequency patterns seems like there is some larger error than only a prefactor/sign :/

**Reply by kinnala** [reply]
I'm not sure I follow. Which of the pictures is H_z? Can you make a minimal example that can be run?

**Reply by HelgeGehring** [reply]
I'll create one

**Reply by HelgeGehring** [reply]
```python
import numpy as np
import matplotlib.pyplot as plt

from skfem import *
from skfem.helpers import *

mesh = MeshTri.init_tensor(x=np.linspace(-1, 1, 20), y=np.linspace(-1, 1, 20))

basis_n0 = Basis(mesh, ElementTriN0() * ElementTriP1())


# e_t(x,y)=[y,-x], e_z = 2
@LinearForm
def from_form(v_t, v_z, w):
    return w.x[1] * v_t[0] - w.x[0] * v_t[1] + 2 * v_z


@BilinearForm
def to_form(e_t, e_z, v_t, v_z, w):
    return inner(e_t, v_t) + inner(e_z, v_z)


x_0 = solve(to_form.assemble(basis_n0), from_form.assemble(basis_n0))
(et, et_basis), (ez, ez_basis) = basis_n0.split(x_0)

et_basis.plot(et)
plt.show()
ez_basis.plot(ez, colorbar=True)
plt.show()


# Calculate

@BilinearForm
def from_form(e_t, e_z, v_t, v_z, w):
    return e_t.curl * v_z + e_z * v_t.curl


@BilinearForm
def to_form(e_t, e_z, v_t, v_z, w):
    return inner(e_t, v_t) + inner(e_z, v_z)


x_1 = solve(to_form.assemble(basis_n0), from_form.assemble(basis_n0) @ x_0)
# bettter x_1 = solve(*condense(to_form.assemble(basis_n0), from_form.assemble(basis_n0) @ x_0, D=basis.get_dofs())) ?
(et, et_basis), (ez, ez_basis) = basis_n0.split(x_1)

et_basis.plot(et)
plt.show()
ez_basis.plot(ez, colorbar=True)
plt.show()

```

I'd expect in the calculation part, that the curl of `e_t` is calculated and stored in `e_z` and that from `e_z` a field with that curl is calculated and stored in `e_t` (not sure about the second part, just added it to understand things better - `e_t.curl` -> `e_z` would be what I need).
So I'd like the arrows to go the other way around and see a value of -2 in e_z after the calculation.

**Reply by HelgeGehring** [reply]
If I use instead of the nedelec element a vector of linear elements, the calculation of e_z seems to work (except at the outer borders of the domain):
```python
import numpy as np
import matplotlib.pyplot as plt

from skfem import *
from skfem.helpers import *

mesh = MeshTri.init_tensor(x=np.linspace(-1, 1, 50), y=np.linspace(-1, 1, 50))

basis = Basis(mesh, ElementVector(ElementTriP1()) * ElementTriP1())


# e_t(x,y)=[y,-x], e_z = 2
@LinearForm
def from_form(v_t, v_z, w):
    return w.x[1] * v_t[0] - w.x[0] * v_t[1] + 2 * v_z


@BilinearForm
def to_form(e_t, e_z, v_t, v_z, w):
    return inner(e_t, v_t) + inner(e_z, v_z)


x_0 = solve(to_form.assemble(basis), from_form.assemble(basis))
(et, et_basis), (ez, ez_basis) = basis.split(x_0)

et_basis.plot(et)
plt.show()
ez_basis.plot(ez, colorbar=True)
plt.show()


# Calculate

@BilinearForm
def from_form(e_t, e_z, v_t, v_z, w):
    return (-e_t.grad[0, 1] + e_t.grad[1, 0]) * v_z + e_z * (-v_t.grad[0, 1] + v_t.grad[1, 0])


@BilinearForm
def to_form(e_t, e_z, v_t, v_z, w):
    return inner(e_t, v_t) + inner(e_z, v_z)


x_1 = solve(*condense(to_form.assemble(basis), from_form.assemble(basis) @ x_0, D=basis.get_dofs()))
(et, et_basis), (ez, ez_basis) = basis.split(x_1)

et_basis.plot(et)
plt.show()
ez_basis.plot(ez, colorbar=True)
plt.show()
```

**Reply by kinnala** [reply]
This should be what you want to do in one line: `ez = ez_basis.project(basis_n0.interpolate(x_0)[0].curl)`

However, now I see that it gives the wrong result. I think we added `ElementTriN0` quite hastily without adding too many tests so perhaps there is a bug lurking somewhere in its implementation.

**Reply by kinnala** [reply]
I might've found the bug. I'll add this as a test to `ElementHcurl` 2D implementation.

**Reply by kinnala** [reply]
I started fixing it in #958. Unfortunately I had to change the rotation direction of 2D curl to be consistent with what we had defined before in `skfem.helpers.curl` for curl of scalar field.

**Reply by HelgeGehring** [reply]
Amazing, thanks! Also nice that the curl will be consistent with the helper function!

Thanks also for the short form, nice to see that it can be written like that :)
Would there be a way to write the projection of the vector-field in the beginning also that short?
```
x_0 = basis.project(lambda x: [[x[1], -x[0]], 2])
```
would be what I've guessed (np.array doesn't work here as it's a nested list), but lists are not accepted there
